substrate/frame/beefy/src/equivocation.rs

```diff
@@ -189,41 +238,77 @@ where
 		reporter: Option<T::AccountId>,
 		evidence: EquivocationEvidenceFor<T>,
 	) -> Result<(), DispatchError> {
-		let (equivocation_proof, key_owner_proof) = evidence;
 		let reporter = reporter.or_else(|| <pallet_authorship::Pallet<T>>::author());
-		let offender = equivocation_proof.offender_id().clone();
+
+		let (offenders, key_owner_proofs, set_id, round) = match &evidence {
+			EquivocationEvidenceFor::VoteEquivocationProof(equivocation_proof, key_owner_proof) =>
+				(
+					vec![equivocation_proof.offender_id()],
+					vec![key_owner_proof.clone()],
+					equivocation_proof.set_id(),
+					*equivocation_proof.round_number(),
+				),
+			EquivocationEvidenceFor::ForkEquivocationProof(
+				equivocation_proof,
+				key_owner_proofs,
+			) => {
+				let offenders = equivocation_proof.offender_ids(); // clone data here
+				(
+					offenders,
+					key_owner_proofs.to_owned(),
+					equivocation_proof.set_id(),
+					*equivocation_proof.round_number(),
+				)
+			},
+		};
 
 		// We check the equivocation within the context of its set id (and
 		// associated session) and round. We also need to know the validator
 		// set count at the time of the offence since it is required to calculate
 		// the slash amount.
-		let set_id = equivocation_proof.set_id();
-		let round = *equivocation_proof.round_number();
-		let session_index = key_owner_proof.session();
-		let validator_set_count = key_owner_proof.validator_count();
-
-		// Validate the key ownership proof extracting the id of the offender.
-		let offender = P::check_proof((BEEFY_KEY_TYPE, offender), key_owner_proof)
-			.ok_or(Error::<T>::InvalidKeyOwnershipProof)?;
-
-		// Validate equivocation proof (check votes are different and signatures are valid).
-		if !sp_consensus_beefy::check_equivocation_proof(&equivocation_proof) {
-			return Err(Error::<T>::InvalidEquivocationProof.into())
-		}
+		let session_index = key_owner_proofs[0].session();
+		let validator_set_count = key_owner_proofs[0].validator_count();
 
 		// Check that the session id for the membership proof is within the
 		// bounds of the set id reported in the equivocation.
-		let set_id_session_index =
-			crate::SetIdSession::<T>::get(set_id).ok_or(Error::<T>::InvalidEquivocationProof)?;
+		let set_id_session_index = crate::SetIdSession::<T>::get(set_id)
+			.ok_or(Error::<T>::InvalidEquivocationProofSession)?;
 		if session_index != set_id_session_index {
-			return Err(Error::<T>::InvalidEquivocationProof.into())
+			return Err(Error::<T>::InvalidEquivocationProofSession.into())
+		}
+
+		// Validate the key ownership proof extracting the ids of the offenders.
+		let offenders = offenders
```

Comment:
Let's move the conversation from https://github.com/paritytech/polkadot-sdk/pull/1329#discussion_r1417060121 here please.

Reply:
> I don't mean for us to handle duplicate offence reports. That's indeed the responsibility of the offences pallet.

> I mean that if someone sends a fork equivocation report with duplicate authorities, signatures, and key ownership proofs, for example: (A, B, C, C, C, C, C ... 1 million times C) and the report is ok, the system will spend compute cycles in checking C 1 million times while the sender will be refunded for the cost of the extrinsic. Which seems like some kind of DoS. Or do we have a limit on the number of authorities in a report somewhere ?

@serban300 I see your point that this would still be a DoS vulnerability.

Do you agree see that this would then be an issue affecting other call sites of the offences pallet too?
If yes, then I'd still suggest this should be fixed in the offences pallet (I reckon in another PR).
A potential solution that comes to mind is to use a `key🡒key_owner_proof` map rather than a vector for storing the `offenders`.

Reply:
> Do you agree that this would then be an issue affecting other call sites of the offences pallet too?
If yes, then I'd still suggest this should be fixed in the offences pallet (I reckon in another PR).

I'm not sure how the offences pallet is used in other places, but for example it wouldn't affect the way it's used from the Grandpa pallet. Because there we only need to send 2 votes for different forks. It seems to be something that depends mostly on the design of the call site. So looks like it would be the responsibility of the call site to handle this.

Reply:
Thanks !

Just one more thing. For the same reason we should also check if the ancestry proof is not bloated (if it doesn't contain unnecessary nodes). I don't know if this is even possible, but leaving it here in order to not forget about it.

@Lederstrumpf do you know if it's possible to have bloated ancestry proofs ? Or if they contain more nodes then necessary, the verification will fail ?